src/util.c

```diff
@@ -552,6 +552,36 @@ int string2d(const char *s, size_t slen, double *dp) {
     return 1;
 }
 
+/* Returns 1 if the double value can safely be represented in long long without
+ * precision loss, in which case the corresponding long long is stored in the out variable. */
+int double2ll(double d, long long *out) {
+#if (DBL_MANT_DIG >= 52) && (DBL_MANT_DIG <= 63) && (LLONG_MAX == 0x7fffffffffffffffLL)
+    /* Check if the float is in a safe range to be casted into a
+     * long long. We are assuming that long long is 64 bit here.
+     * Also we are assuming that there are no implementations around where
+     * double has precision < 52 bit.
+     *
+     * Under this assumptions we test if a double is inside a range
+     * where casting to long long is safe. Then using two castings we
+     * make sure the decimal part is zero. If all this is true we can use
+     * integer without precision loss.
+     *
+     * Note that numbers above 2^52 and below 2^63 use all the fraction bits as real part,
+     * and the exponent bits are positive, which means the "decimal" part must be 0.
+     * i.e. all double values in that range are representable as a long without precision loss,
+     * but not all long values in that range can be represented as a double.
+     * we only care about the first part here. */
+    if (d < -LLONG_MAX || d > LLONG_MAX)
```

Comment:
Here is an implicit conversion, LLONG_MAX will be converted to double, it will be 9223372036854775808.
https://godbolt.org/z/zs5zaoqv8

Reply:
Second thoughts, it doesn't seem to affect, in the following `if (ll == d)` it will be false.

Reply:
the conversion existed before.
```c
double max = 4503599627370496;
```
i was counting on it (to implicitly happen).
the problem you mention (falsely thinking that we're in range) is due to the precision loss, we can fix it by reducing the range by half.
https://godbolt.org/z/EdnMn1qoo